pkg/component/controller/workerconfig/reconciler.go

```diff
@@ -0,0 +1,606 @@
+
+type reconcilerStarted struct {
+	stop    func()
+	stopped <-chan struct{}
+
+	updates chan<- updateFunc
+}
+
+type reconcilerCreated struct {
+	clientFactory kubeutil.ClientFactoryInterface
+}
+
+// NewReconciler creates a new reconciler for worker configurations.
+func NewReconciler(k0sVars constant.CfgVars, nodeSpec *v1beta1.ClusterSpec, clientFactory kubeutil.ClientFactoryInterface, leaderElector leaderelector.Interface) (*Reconciler, error) {
+	log := logrus.WithFields(logrus.Fields{"component": "workerconfig.Reconciler"})
+
+	clusterDNSIPString, err := nodeSpec.Network.DNSAddress()
+	if err != nil {
+		return nil, err
+	}
+	clusterDNSIP := net.ParseIP(clusterDNSIPString)
+	if clusterDNSIP == nil {
+		return nil, fmt.Errorf("not an IP address: %q", clusterDNSIPString)
+	}
+
+	reconciler := &Reconciler{
+		log: log,
+
+		leaderElector: leaderElector,
+		clusterDomain: nodeSpec.Network.ClusterDomain,
+		clusterDNSIP:  clusterDNSIP,
+		cleaner: &kubeletConfigCleaner{
+			log: log,
+
+			dir:           filepath.Join(k0sVars.ManifestsDir, "kubelet"),
+			clientFactory: clientFactory,
+		},
+		state: reconcilerCreated{clientFactory},
+	}
+
+	return reconciler, nil
+}
+
+type reconcilerInitialized struct {
+	client kubernetes.Interface
+	apply  func(context.Context, resources) error
+}
+
+func (r *Reconciler) Init(context.Context) error {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	var created reconcilerCreated
+	switch state := r.state.(type) {
+	case reconcilerCreated:
+		created = state
+	default:
+		return fmt.Errorf("cannot initialize: %T", state)
+	}
+
+	client, err := created.clientFactory.GetClient()
+	if err != nil {
+		return err
+	}
+
+	apply := func(ctx context.Context, resources resources) error {
+		dynamicClient, err := created.clientFactory.GetDynamicClient()
+		if err != nil {
+			return err
+		}
+		discoveryClient, err := created.clientFactory.GetDiscoveryClient()
+		if err != nil {
+			return err
+		}
+
+		return (&applier.Stack{
+			Name:      "k0s-" + constant.WorkerConfigComponentName,
+			Client:    dynamicClient,
+			Discovery: discoveryClient,
+			Resources: resources,
+		}).Apply(ctx, true)
+	}
+
+	r.cleaner.init()
+	r.state = reconcilerInitialized{client, apply}
+	return nil
+}
+
+func (r *Reconciler) Start(context.Context) error {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	var initialized reconcilerInitialized
+	switch state := r.state.(type) {
+	case reconcilerInitialized:
+		initialized = state
+	default:
+		return fmt.Errorf("cannot start: %T", state)
+	}
+
+	var started reconcilerStarted
+	updates := make(chan updateFunc, 1)
+	started.updates = updates
+
+	reconcilersCtx, cancelReconcilers := context.WithCancel(context.Background())
+	var reconcilersDone sync.WaitGroup
+
+	// Will be locked during callbacks of the leader elector.
+	// TODO: Replace this with some "callback removal" procedure, once available.
+	var leaderElectorMu sync.Mutex
+
+	// Set to true once the reconciler gets stopped.
+	var stopCalled atomic.Bool
```

Comment:
Well, three different async patterns in a row make this code really complicated. Is there any way to simplify it?

Reply:
I think I found a better solution. Will update the PR accordingly.

Reply:
Done. No more atomics. Please have another look.